flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1155,13 +1155,22 @@ mlir::Value IntrinsicLibrary::genIchar(mlir::Type resultType,
   assert(args.size() >= 1);
 
   auto arg = args[0];
+  assert(Fortran::lower::CharacterExprHelper::isCharacter(arg.getType()) &&
+         "Error: Unhandled type passed to ICHAR");
   Fortran::lower::CharacterExprHelper helper{builder, loc};
-  auto dataAndLen = helper.createUnboxChar(arg);
```

Comment:
Is this going to work with a boxchar or a character with a dynamic LEN type parameter?